File: 백준/Gold/1167. 트리의 지름/트리의 지름.cc
#include<bits/stdc++.h>
using namespace std;
int V,max_cost=0,max_node;
vector<vector<pair<int,int>>>edges;
vector<bool>visited;

void dfs(int cur_node,int cur_cost){
    if(cur_cost>max_cost){
        max_node=cur_node;
        max_cost=cur_cost;
    }
    for(size_t i=0;i<edges[cur_node].size();i++){
        int next_node=edges[cur_node][i].first;
        int next_cost=edges[cur_node][i].second+cur_cost;
        if(!visited[next_node]){
            visited[next_node]=true;
            dfs(next_node,next_cost);
            visited[next_node]=false;
        }
    }
}

int main(){
    int s,e,c;
    cin>>V;
    edges.resize(V+1);
    for(int i=0;i<V;i++){
        cin>>s;
        while(true){
            cin>>e;
            if(e==-1){
                break;
            }
            cin>>c;
            edges[s].push_back({e,c});
        }
    }

    // for(size_t i=0;i<edges[4].size();i++){
    //     cout<<edges[4][i].first<<" "<<edges[4][i].second<<"\n";
    // }

    visited.resize(V+1,false);
    visited[1]=true;
    dfs(1,0);
    int tmp=max_node;
    // cout<<tmp<<" "<<max_cost<<"\n";

    max_cost=0;
    fill(visited.begin(),visited.end(),false);
    visited[tmp]=true;
    dfs(tmp,0);
    cout<<max_cost;
    return 0;
}

File: 백준/Gold/1167. 트리의 지름/트리의 지름.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	static int V, max_cost = 0, max_node = 0;
	static ArrayList<Node>[] edges;
	static int[] dist;
	static boolean[] visited;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();

		StringTokenizer st = new StringTokenizer(br.readLine(), " ");
		V = Integer.parseInt(st.nextToken());

		edges = new ArrayList[V + 1];
		for (int i = 0; i <= V; i++) {
			edges[i] = new ArrayList<Node>();
		}
		visited = new boolean[V + 1];
		dist = new int[V + 1];

		int start, end, cost;
		for (int i = 0; i < V; i++) {
			st = new StringTokenizer(br.readLine());
			start = Integer.parseInt(st.nextToken());
			while (true) {
				end = Integer.parseInt(st.nextToken());
				if (end == -1) {
					break;
				}
				cost = Integer.parseInt(st.nextToken());
				edges[start].add(new Node(end, cost));
			}
		}
		bfs(1);
		max_cost = 0;
		visited = new boolean[V + 1];
		dist = new int[V + 1];
		bfs(max_node);
		System.out.println(max_cost);
	}

	public static void bfs(int start) {
		Queue<Node> q = new ArrayDeque<>();
		q.add(new Node(start, 0));
		dist[start] = 0;
		visited[start] = true;
		while (!q.isEmpty()) {
			Node cur = q.poll();
			for (int i = 0; i < edges[cur.end].size(); i++) {
				int next_node = edges[cur.end].get(i).end;
				int next_cost = edges[cur.end].get(i).cost + cur.cost;
				if(visited[next_node])
					continue;
				visited[next_node] = true;
				dist[next_node] = next_cost;
				q.add(new Node(next_node, next_cost));
			}
		}
		for (int i = 1; i <= V; i++) {
			if (max_cost < dist[i]) {
				max_node = i;
				max_cost = dist[i];
			}
		}
	}

	public static class Node {
		int end, cost;

		public Node(int end, int cost) {
			this.end = end;
			this.cost = cost;
		}
	}
}